### ISSUE TITLE
nn.Softmax should not allow default/implicit/unset dim constructor argument

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#84524]
### 🐛 Describe the bug

Originally discussed in https://github.com/pytorch/pytorch/issues/84290#issuecomment-1232133690

The F.softmax / softmin / log_softmax have implicit dim support deprecated since a few years ago, so nn.Softmax should also remove this support for reducing possible confusion

cc @alband @jerrryzh

### Versions

N/A

**Comment by zou3519** [comment]
Sounds reasonable

**Comment by vadimkantorov** [comment]
Would be good to also carry out old-decided deprecation of implicit dim for underlying methods: https://github.com/pytorch/pytorch/issues/36524

**Comment by albanD** [comment]
This might be ok to do but I think we need to look at downstream libs to see how disruptive this will be first.

**Comment by jbschlosser** [comment]
Hey @kit1980, I see you're assigned here - are you interesting in digging into this some (e.g. doing a GitHub regex code search to see the impact of requiring the `dim` arg)?
